executor/write.go

```diff
@@ -788,28 +790,33 @@ type InsertExec struct {
 	Priority  mysql.PriorityEnum
 	IgnoreErr bool
 
-	finished bool
+	finished      bool
+	rowCount      int
+	dirtyRowValue map[string][]byte
 }
 
 func (e *InsertExec) exec(ctx context.Context, rows [][]types.Datum) (Row, error) {
 	// If tidb_batch_insert is ON and not in a transaction, we could use BatchInsert mode.
 	sessVars := e.ctx.GetSessionVars()
 	defer sessVars.CleanBuffers()
-	batchInsert := sessVars.BatchInsert && !sessVars.InTxn()
-	batchSize := sessVars.DMLBatchSize
 
-	txn := e.ctx.Txn()
-	rowCount := 0
-	writeBufs := sessVars.GetWriteStmtBufs()
+	e.rowCount = 0
 	if !sessVars.ImportingData {
-		writeBufs.BufStore = kv.NewBufferStore(txn, kv.TempTxnMemBufCap)
+		sessVars.GetWriteStmtBufs().BufStore = kv.NewBufferStore(e.ctx.Txn(), kv.TempTxnMemBufCap)
+	}
+	if len(e.OnDuplicate) > 0 && !e.IgnoreErr {
+		var err error
+		rows, err = e.batchUpdateDupRows(rows, e.OnDuplicate)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
 	}
 	// If you use the IGNORE keyword, duplicate-key error that occurs while executing the INSERT statement are ignored.
 	// For example, without IGNORE, a row that duplicates an existing UNIQUE index or PRIMARY KEY value in
 	// the table causes a duplicate-key error and the statement is aborted. With IGNORE, the row is discarded and no error occurs.
 	// However, if the `on duplicate update` is also specified, the duplicated row will be updated.
 	// Using BatchGet in insert ignore to mark rows as duplicated before we add records to the table.
-	if e.IgnoreErr && len(e.OnDuplicate) == 0 {
+	if len(e.OnDuplicate) == 0 && e.IgnoreErr {
```

Comment:
What is difference in this change? To me they are exactly same.

Reply:
It means nothing, but the insert behavior should always check duplicate key update first. Just keep them in a consistent order.

Reply:
gotcha.